File: CSharp/138-CopyListWithRandomPointer/138CopyListWithRandomPointer.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;
    
    public Node(int _val) {
        val = _val;
        next = null;
        random = null;
    }
}
*/

// HASH MAP && TWO PASSES

// Time Complexity: O(n)
// each node is visited twice times but it's still linear (2n reduced to On?)

// Space Complexity: O(n)
// Each node has to be stored in the hash map. 
public class Solution {
    public Node CopyRandomList(Node head) {
     if (head == null) return null;
     Dictionary<Node, Node> oldToNew = new Dictionary<Node,Node>();
     Node curr = head;
     while (curr != null){
        oldToNew[curr] = new Node(curr.val);
        curr = curr.next;
     }   
     curr = head;
     while (curr != null){
        oldToNew[curr].next = curr.next != null ? oldToNew[curr.next] : null;
        oldToNew[curr].random = curr.random != null ? oldToNew[curr.random] : null;
        curr = curr.next;
     }
     return oldToNew[head];
    }
}

File: Javascript/138-CopyListWithRandomPointer/138CopyListWithRandomPointer.js
/* Hash Map
Time Complexity: O(n)
* Each node is visited twice, so O(2n) which simplifies to O(n).
Space Complexity: O(n)
* We have to store all nodes in the hash map. 
*/
/************************************************
 * // Definition for a _Node.
 * function _Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {_Node} head
 * @return {_Node}
 */
var copyRandomList = function(head) {
    if (!head) return null;
    const oldToNew = new Map();
    let curr = head;
    while (curr) {
        oldToNew.set(curr, new Node(curr.val));
        curr = curr.next;
    }
    curr = head;
    while (curr){
        oldToNew.get(curr).next = oldToNew.get(curr.next) || null;
        oldToNew.get(curr).random = oldToNew.get(curr.random) || null;
        curr = curr.next;
    }
    return oldToNew.get(head);
};